contracts/cw721-base/src/contract.rs

```diff
@@ -0,0 +1,499 @@
+}
+
+pub fn handle<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    msg: HandleMsg,
+) -> StdResult<HandleResponse> {
+    match msg {
+        HandleMsg::Mint {
+            token_id,
+            owner,
+            name,
+            description,
+            image,
+        } => handle_mint(deps, env, token_id, owner, name, description, image),
+        HandleMsg::Approve {
+            spender,
+            token_id,
+            expires,
+        } => handle_approve(deps, env, spender, token_id, expires),
+        HandleMsg::Revoke { spender, token_id } => handle_revoke(deps, env, spender, token_id),
+        HandleMsg::ApproveAll { operator, expires } => {
+            handle_approve_all(deps, env, operator, expires)
+        }
+        HandleMsg::RevokeAll { operator } => handle_revoke_all(deps, env, operator),
+        HandleMsg::TransferNft {
+            recipient,
+            token_id,
+        } => handle_transfer_nft(deps, env, recipient, token_id),
+        HandleMsg::SendNft {
+            contract,
+            token_id,
+            msg,
+        } => handle_send_nft(deps, env, contract, token_id, msg),
+    }
+}
+
+pub fn handle_mint<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    token_id: String,
+    owner: HumanAddr,
+    name: String,
+    description: Option<String>,
+    image: Option<String>,
+) -> StdResult<HandleResponse> {
+    let minter = mint(&mut deps.storage).load()?;
+    let minter_human = deps.api.human_address(&minter)?;
+
+    if minter_human != env.message.sender {
+        return Err(StdError::unauthorized());
+    }
+
+    // create the token
+    let token = TokenInfo {
+        owner: deps.api.canonical_address(&owner)?,
+        approvals: vec![],
+        name,
+        description: description.unwrap_or_default(),
+        image,
+    };
+    tokens(&mut deps.storage).save(token_id.as_bytes(), &token)?;
+
+    increment_tokens(&mut deps.storage)?;
+
+    // TODO: set logs
+    Ok(HandleResponse::default())
+}
+
+pub fn handle_transfer_nft<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    recipient: HumanAddr,
+    token_id: String,
+) -> StdResult<HandleResponse> {
+    let mut token = tokens(&mut deps.storage).load(token_id.as_bytes())?;
+    // ensure we have permissions
+    check_can_send(&deps, &env, &token)?;
+    // set owner and remove existing approvals
+    token.owner = deps.api.canonical_address(&recipient)?;
+    token.approvals = vec![];
+    tokens(&mut deps.storage).save(token_id.as_bytes(), &token)?;
+
+    // TODO: set logs
+    Ok(HandleResponse::default())
+}
+
+pub fn handle_send_nft<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    contract: HumanAddr,
+    token_id: String,
+    _msg: Option<Binary>,
+) -> StdResult<HandleResponse> {
+    let mut token = tokens(&mut deps.storage).load(token_id.as_bytes())?;
```

Comment:
This is just `handle_transfer_nft()` (plus the send msg), and can probably be simplified as such.